### PR TITLE
GSYE-111: Add 'matching_requirements' field to the BidOfferMatch

### PR DESCRIPTION
**Pull request by Hassan754**
The orders could have a list of energy requirements, when creating recommendations only one among the requirements is adopted.
The newly added matching requirements at the moment keeps track of the adopted requirement dictionary, and is scalable enough to allow the client to send any argument to GSYE (for researches who may need to send custom variables to the backend)

### CODE REVIEW

File: gsy_framework/data_classes.py [hunk 1]
```diff
@@ -487,18 +489,48 @@ def from_dict(cls, bid_offer_match: Dict) -> Optional["BidOfferMatch"]:
...
             is_valid = False
         elif not isinstance(bid_offer_match["trade_rate"], (int, float)):
             is_valid = False
+        elif not (bid_offer_match.get("matching_requirements") is None or
+                  isinstance(bid_offer_match.get("matching_requirements"), Dict)):
+            is_valid = False
         return is_valid
 
+    @property
+    def bid_energy(self):
```
Comment (Hassan754) [inline gsy_framework/data_classes.py:508]: @spyrostz can you check this out, if this approach is okay I will move on with it

Comment (spyrostz) [reply gsy_framework/data_classes.py:508]: @Hassan754 looks good to me , and it is flexible enough to add more keys in the table if such a need arises in the future!

**Review by spyrostz — APPROVED** [review]
LGTM

**Review by hannesdiedrich — DISMISSED** [review]
LGTM

**Review by hannesdiedrich — APPROVED** [review]

**Review by spyrostz — APPROVED** [review]
LGTM

**Pull request merged**